File: Anul II/Medii de proiectare si programare/Laborator/lab08/C#/Ubb.BikeContest.Networking/ProtocolBuffers/ProtocolBufferWorker.cs
using Com.Ubb.Protocolbuffers;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;
using Ubb.BikeContest.Model;
using Ubb.BikeContest.Model.data;
using Ubb.BikeContest.Services;

using User = Ubb.BikeContest.Model.User;
using Participant = Ubb.BikeContest.Model.Participant;
using Race = Ubb.BikeContest.Model.Race;
using RaceEntry = Ubb.BikeContest.Model.RaceEntry;
using Team = Ubb.BikeContest.Model.Team;
using RaceDTO = Ubb.BikeContest.Model.data.RaceDto;
using Google.Protobuf;

namespace Ubb.BikeContest.Networking.ProtocolBuffers
{
    public class ProtocolBufferWorker : IMainObserver
    {
        private readonly IContestServices _server;
        private readonly TcpClient _connection;
        private readonly NetworkStream _stream;
        private volatile bool _connected;

        public ProtocolBufferWorker(IContestServices server, TcpClient connection)
        {
            _server = server;
            _connection = connection;

            try
            {
                _stream = _connection.GetStream();
                _connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void ParticipantAdded(Participant participant)
        {
            try
            {
                SendResponse(ProtocolBuilderUtils.CreateNewParticipantResponse(participant));
            }
            catch (Exception e)
            {
                Console.WriteLine("Error while sending response object: " + e.Message);
            }
        }

        public virtual void RaceEntriesAdded(List<RaceDto> races)
        {
            try
            {
                SendResponse(ProtocolBuilderUtils.CreateUpdatedRacesResponse(races));
            }
            catch (Exception e)
            {
                Console.WriteLine("Error while sending response object: " + e.Message);
            }
        }

        private BikeContestResponse HandleRequest(BikeContestRequest request)
        {
            BikeContestRequest.Types.Type type = request.Type;

            switch (type)
            {
                case BikeContestRequest.Types.Type.Login:
                    Console.WriteLine("> handling login request...");
                    string username = ProtocolBuilderUtils.GetUsername(request);
                    string passwordToken = ProtocolBuilderUtils.GetPasswordToken(request);

                    try
                    {
                        lock (_server)
                        {
                            User connectedUser = _server.Login(username, passwordToken, this);
                            return ProtocolBuilderUtils.CreateOkResponse(connectedUser);
                        }
                    }
                    catch (Exception e)
                    {
                        _connected = false;
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.Logout:
                    Console.WriteLine("> handling logout request...");
                    User user = ProtocolBuilderUtils.GetUser(request);

                    try
                    {
                        lock (_server)
                        {
                            _server.Logout(user, this);
                        }

                        _connected = false;
                        return ProtocolBuilderUtils.CreateOkResponse();
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.CreateParticipant:
                    Console.WriteLine("> handling create participant request...");
                    Participant participant = ProtocolBuilderUtils.GetParticipant(request);

                    try
                    {
                        lock (_server)
                        {
                            _server.SaveParticipant(participant);

                            return ProtocolBuilderUtils.CreateNewParticipantResponse(
                                _server.GetParticipantByData(participant));
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.CreateRace:
                    Console.WriteLine("> handling create race request...");
                    Race race = ProtocolBuilderUtils.GetRace(request);

                    try
                    {
                        lock (_server)
                        {
                            _server.SaveRace(race);
                            Race createdRace = _server.GetRaceByName(race.Name);
                            List<RaceEntry> entries = _server.GetEntriesByRace(createdRace.Id);

                            return ProtocolBuilderUtils.CreateNewRaceResponse(
                                new RaceDTO(createdRace.Name, createdRace.EngineCapacity, entries.Count));
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.CreateRaceEntries:
                    Console.WriteLine("> handling create race entries request...");
                    List<RaceEntry> raceEntries = ProtocolBuilderUtils.GetRaceEntries(request);

                    try
                    {
                        lock (_server)
                        {
                            _server.SaveRaceEntries(raceEntries);

                            return ProtocolBuilderUtils.CreateUpdatedRacesResponse(
                                _server.GetRacesWithParticipantCount());
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.GetEntriesByRace:
                    Console.WriteLine("> handling get entries by race request...");
                    long raceId = ProtocolBuilderUtils.GetRaceId(request);

                    try
                    {
                        lock (_server)
                        {
                            List<RaceEntry> entries = _server.GetEntriesByRace(raceId);
                            return ProtocolBuilderUtils.CreateEntriesByRaceResponse(entries);
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.GetParticipantByData:
                    Console.WriteLine("> handling get participant by data request...");
                    Participant data = ProtocolBuilderUtils.GetParticipant(request);

                    try
                    {
                        lock (_server)
                        {
                            return ProtocolBuilderUtils.CreateParticipantResponse(
                                _server.GetParticipantByData(data));
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.GetParticipantsByTeam:
                    Console.WriteLine("> handling get participants by team request...");
                    long teamId = ProtocolBuilderUtils.GetTeamId(request);

                    try
                    {
                        lock (_server)
                        {
                            List<Participant> participants = _server.GetParticipantsByTeam(teamId);
                            return ProtocolBuilderUtils.CreateParticipantsByTeamResponse(participants);
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.GetParticipants:
                    Console.WriteLine("> handling get participants request...");

                    try
                    {
                        lock (_server)
                        {
                            List<Participant> participants = _server.FindAllParticipants();

                            return ProtocolBuilderUtils.CreateAllParticipantsResponse(participants);
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.GetRaceByName:
                    Console.WriteLine("> handling get race by name request...");
                    string raceName = ProtocolBuilderUtils.GetRaceName(request);

                    try
                    {
                        lock (_server)
                        {
                            Race res = _server.GetRaceByName(raceName);

                            return ProtocolBuilderUtils.CreateGetRaceByNameResponse(res);
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.GetRacesByEngineCapacity:
                    Console.WriteLine("> handling get races by engine capacity request...");
                    int capacity = ProtocolBuilderUtils.GetEngineCapacity(request);

                    try
                    {
                        lock (_server)
                        {
                            List<Race> races = _server.GetRacesByEngineCapacity(capacity);

                            return ProtocolBuilderUtils.CreateGetRacesByCapacityResponse(races);
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.GetRacesWithParticipantCount:
                    Console.WriteLine("> handling get races with participant count request...");

                    try
                    {
                        lock (_server)
                        {
                            List<RaceDTO> races = _server.GetRacesWithParticipantCount();

                            return ProtocolBuilderUtils.CreateGetRacesWithParticipantsResponse(races);
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.GetTeamByName:
                    Console.WriteLine("> handling get team by name request...");
                    string teamName = ProtocolBuilderUtils.GetTeamName(request);

                    try
                    {
                        lock (_server)
                        {
                            Team team = _server.GetTeamByName(teamName);

                            return ProtocolBuilderUtils.CreateGetTeamResponse(team);
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.GetTeams:
                    Console.WriteLine("> handling get teams request...");

                    try
                    {
                        lock (_server)
                        {
                            List<Team> teams = _server.FindAllTeams();

                            return ProtocolBuilderUtils.CreateAllTeamsResponse(teams);
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case BikeContestRequest.Types.Type.GetUnregisteredRaces:
                    Console.WriteLine("> handling get unregistered races request...");
                    long participandId = ProtocolBuilderUtils.GetParticipantId(request);
                    int engineCapacity = ProtocolBuilderUtils.GetEngineCapacity(request);

                    try
                    {
                        lock (_server)
                        {
                            List<Race> races = _server.GetRacesWhereNotRegisteredAndEngineCapacity(participandId, engineCapacity);

                            return ProtocolBuilderUtils.CreateGetRacesByCapacityResponse(races);
                        }
                    }
                    catch (Exception e)
                    {
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }
            }
            return null;
        }

        private void SendResponse(BikeContestResponse response)
        {
            Console.WriteLine("Sending response: " + response);

            lock (_stream)
            {
                response.WriteDelimitedTo(_stream);
                _stream.Flush();
            }
        }

        public virtual void Run()
        {
            while (_connected)
            {
                try
                {
                    BikeContestRequest request = BikeContestRequest.Parser.ParseDelimitedFrom(_stream);
                    BikeContestResponse response = HandleRequest(request);

                    if (response != null)
                    {
                        SendResponse(response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }

            try
            {
                _stream.Close();
                _connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }
    }
}

File: Anul II/Medii de proiectare si programare/Laborator/lab11/java/networking/src/main/java/com/ubb/protocolbuffers/ProtocolBufferWorker.java
package com.ubb.protocolbuffers;

import com.ubb.IContestServices;
import com.ubb.IMainObserver;
import com.ubb.dto.RaceDTO;
import com.ubb.exceptions.ContestDataException;
import com.ubb.model.*;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class ProtocolBufferWorker implements Runnable, IMainObserver {
    private final IContestServices server;
    private final Socket connection;
    private InputStream inputStream;
    private OutputStream outputStream;
    private volatile boolean connected;

    public ProtocolBufferWorker(IContestServices server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try {
            outputStream = connection.getOutputStream();
            inputStream = connection.getInputStream();
        } catch (IOException ioException) {
            ioException.printStackTrace();
        }
    }

    @Override
    public void raceAdded(RaceDTO race) throws ContestDataException {
        try {
            BikeContestProtocol.BikeContestResponse response = ProtocolBuilderUtils.createNewRaceResponse(race);
            System.out.println("Race added: " + race);
            sendResponse(response);
        } catch (IOException ioException) {
            throw new ContestDataException("Error while sending response object: " + ioException);
        }
    }

    @Override
    public void participantAdded(Participant participant) throws ContestDataException {
        try {
            BikeContestProtocol.BikeContestResponse response
                    = ProtocolBuilderUtils.createNewParticipantResponse(participant);
            System.out.println("Participant added: " + participant);
            sendResponse(response);
        } catch (IOException ioException) {
            throw new ContestDataException("Error while sending response object: " + ioException);
        }
    }

    @Override
    public void raceEntriesAdded(List<RaceDTO> races) throws ContestDataException {
        try {
            BikeContestProtocol.BikeContestResponse response
                    = ProtocolBuilderUtils.createUpdatedRacesResponse(races);
            System.out.println("Race entries added: " + races);
            sendResponse(response);
        } catch (IOException ioException) {
            throw new ContestDataException("Error while sending response object: " + ioException);
        }
    }

    @Override
    public void run() {
        while (connected) {
            try {
                System.out.println("Waiting for requests...");
                BikeContestProtocol.BikeContestRequest request
                        = BikeContestProtocol.BikeContestRequest.parseDelimitedFrom(inputStream);
                System.out.println("Request received: " + request);

                BikeContestProtocol.BikeContestResponse response = handleRequest(request);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (IOException ioException) {
                ioException.printStackTrace();
            }

            try {
                Thread.sleep(1000);
            } catch (InterruptedException interruptedException) {
                interruptedException.printStackTrace();
            }
        }
    }

    private BikeContestProtocol.BikeContestResponse handleRequest(BikeContestProtocol.BikeContestRequest request) {
        BikeContestProtocol.BikeContestResponse response = null;

        switch (request.getType()) {
            case Login -> {
                System.out.println("> handling login request...");
                final String username = ProtocolBuilderUtils.getUsername(request);
                final String passwordToken = ProtocolBuilderUtils.getPasswordToken(request);

                try {
                    final User loggedUser = server.login(username, passwordToken, this);
                    return ProtocolBuilderUtils.createOkResponse(loggedUser);
                } catch (ContestDataException contestDataException) {
                    connected = false;
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case Logout -> {
                System.out.println("> handling logout request");
                final User user = ProtocolBuilderUtils.getUser(request);

                try {
                    server.logout(user, this);
                    connected = false;
                    return ProtocolBuilderUtils.createOkResponse();
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case CreateRace -> {
                System.out.println("> handling create race request...");
                final Race race = ProtocolBuilderUtils.getRaceDto(request);

                try {
                    server.saveRace(race);
                    final Optional<Race> createdRace = server.getRaceByName(race.getName());
                    if (createdRace.isEmpty()) {
                        throw new ContestDataException("Race could not be created.");
                    }
                    List<RaceEntry> raceEntries = new ArrayList<>(server.getEntriesByRace(createdRace.get().getId()));

                    return ProtocolBuilderUtils.createNewRaceResponse(
                            new RaceDTO(race.getName(), race.getEngineCapacity(), raceEntries.size()));
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case CreateParticipant -> {
                System.out.println("> handling create participant request");
                final Participant participant = ProtocolBuilderUtils.getParticipant(request);

                try {
                    server.saveParticipant(participant);
                    final Optional<Participant> savedParticipant = server.getParticipantByData(participant);
                    if (savedParticipant.isEmpty()) {
                        throw new ContestDataException("Participant could not be created.");
                    }
                    return ProtocolBuilderUtils.createNewParticipantResponse(savedParticipant.get());
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case CreateRaceEntries -> {
                System.out.println("> handling create race entries request");
                final List<RaceEntry> raceEntries = ProtocolBuilderUtils.getRaceEntries(request);

                try {
                    server.saveRaceEntries(raceEntries);

                    return ProtocolBuilderUtils.createUpdatedRacesResponse(
                            server.getRacesWithParticipantCount().stream().toList());
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case GetEntriesByRace -> {
                System.out.println("> handling get entries by race request");
                final Long raceId = ProtocolBuilderUtils.getRaceId(request);

                try {
                    List<RaceEntry> raceEntries = server.getEntriesByRace(raceId).stream().toList();

                    return ProtocolBuilderUtils.createEntriesByRaceResponse(raceEntries);
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case GetParticipantByData -> {
                System.out.println("> handling get participant by data request");
                final Participant participant = ProtocolBuilderUtils.getParticipant(request);

                try {
                    Optional<Participant> found = server.getParticipantByData(participant);
                    if (found.isEmpty()) {
                        throw new ContestDataException("Participant could not be found.");
                    }
                    return ProtocolBuilderUtils.createParticipantResponse(found.get());
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case GetParticipantsByTeam -> {
                System.out.println("> handling get participants by team request");
                final Long teamId = ProtocolBuilderUtils.getTeamId(request);

                try {
                    List<Participant> participants = server.getParticipantsByTeam(teamId).stream().toList();

                    return ProtocolBuilderUtils.createParticipantsByTeamResponse(participants);
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case GetParticipants -> {
                System.out.println("> handling get participants request");

                try {
                    List<Participant> participants = server.findAllParticipants().stream().toList();

                    return ProtocolBuilderUtils.createAllParticipantsResponse(participants);
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case GetRaceByName -> {
                System.out.println("> handling get race by name request");
                final String raceName = ProtocolBuilderUtils.getRaceName(request);

                try {
                    Optional<Race> found = server.getRaceByName(raceName);
                    if (found.isEmpty()) {
                        throw new ContestDataException("Participant could not be found.");
                    }

                    return ProtocolBuilderUtils.createGetRaceByNameResponse(found.get());
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case GetRacesByEngineCapacity -> {
                System.out.println("> handling get races by capacity request");
                final Integer capacity = ProtocolBuilderUtils.getEngineCapacity(request);

                try {
                    List<Race> races = server.getRacesByEngineCapacity(capacity).stream().toList();

                    return ProtocolBuilderUtils.createGetRacesByCapacityResponse(races);
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case GetRacesWithParticipantCount -> {
                System.out.println("> handling get races with participants request");

                try {
                    List<RaceDTO> races = server.getRacesWithParticipantCount().stream().toList();

                    return ProtocolBuilderUtils.createGetRacesWithParticipantsResponse(races);
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case GetTeamByName -> {
                System.out.println("> handling get team by name request");
                final String teamName = ProtocolBuilderUtils.getTeamName(request);

                try {
                    Optional<Team> found = server.getTeamByName(teamName);
                    if (found.isEmpty()) {
                        throw new ContestDataException("Participant could not be found.");
                    }

                    return ProtocolBuilderUtils.createGetTeamResponse(found.get());
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case GetTeams -> {
                System.out.println("> handling get teams request");

                try {
                    List<Team> teams = server.findAllTeams().stream().toList();

                    return ProtocolBuilderUtils.createAllTeamsResponse(teams);
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case GetUnregisteredRaces -> {
                System.out.println("> handling get unregistered races request");
                final Long participantId = ProtocolBuilderUtils.getParticipantId(request);
                final Integer capacity = ProtocolBuilderUtils.getEngineCapacity(request);

                try {
                    List<Race> races
                            = server.getRacesWhereNotRegisteredAndEngineCapacity(participantId, capacity)
                            .stream().toList();

                    return ProtocolBuilderUtils.createGetRacesByCapacityResponse(races);
                } catch (ContestDataException contestDataException) {
                    return ProtocolBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }
        }

        return null;
    }

    private void sendResponse(BikeContestProtocol.BikeContestResponse response) throws IOException {
        System.out.println("Sending response: " + response);
        response.writeDelimitedTo(outputStream);
        outputStream.flush();
    }
}
